plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java

```diff
@@ -2203,7 +2203,7 @@ protected StartAnswer execute(StartCommand cmd) {
                     throw new Exception("Failed to find the newly create or relocated VM. vmName: " + vmInternalCSName);
                 }
             }
-            if (deployAsIs) {
+            if (deployAsIs && !vmMo.hasSnapshot()) {
```

Comment:
This seems right to me as I see some other calls of tearDownVmDevices() method before this call which does handle snapshot existance.
https://github.com/shapeblue/cloudstack/blob/78e07cff62dd98d9b88852a47407e80fa57537d2/plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java#L2826-L2834

    private void tearDownVmDevices(VirtualMachineMO vmMo, boolean hasSnapshot, boolean deployAsIs) throws Exception {
        if (deployAsIs) {
            vmMo.tearDownDevices(new Class<?>[]{VirtualEthernetCard.class});
        } else if (!hasSnapshot) {
            vmMo.tearDownDevices(new Class<?>[]{VirtualDisk.class, VirtualEthernetCard.class});
        } else {
            vmMo.tearDownDevices(new Class<?>[]{VirtualEthernetCard.class});
        }
    }

Reply:
thanks @harikrishna-patnala 
It looks consistent with this pr.

I got the error on vcenter: " Invalid configuration for device '0'. Cannot remove virtual disk from the virtual machine because it or one of its parent disks is part of a snapshot of the virtual machine."
it seems to be a known restriction that disk cannot be removed from vm with snapshot.